csrc/expr_evaluator.cpp

```diff
@@ -141,6 +141,11 @@ void ExpressionEvaluator::bind_(
             i);
         bind_(
             rfactor_domain[i]->expandedExtent(), t.size(i), evaluate_validate);
+      } else if (rfactor_domain[i]->isDeviceDim()) {
+        // Device dimensions extents will always be 1.
+        // Ignore concrete extents because they hold the unsharded extents.
```

Comment:
I'm not sure if we should just always ignore this case. At this point, we should know that `rfactor_domain[i]->extent()` is equal to the extent of `DIDx`.

Reply:
DIDx holds the unsharded extent. This case handles global sharded inputs: if an axis is sharded then we ignore the unsharded extent and check that it's size 1.

Reply:
Right, so we know the unsharded extent at this point, and that should be equal to `rfactor_domain[i]->extent()`, right?

Reply:
Well, actually, that isn't guaranteed. The only thing we can say is: `number-of-devices-in-DIDx >= rfactor_domain[i]->extent()`. So, I don't think we can bind anything in this case.

Reply:
This is the binding gets triggered when we have a sharded concrete tensorview input. We create the tensorview with the unsharded extents, so the unsharded extent is known. If the axis is sharded then the tensor's axis should be 1 otherwise it should be the unsharded size.

Since we don't handle split yet `number-of-devices-in-DIDx == rfactor_domain[i]->extent()`

Reply:
Ahh if that is the case we have a much bigger problem since all the multidevice logic currently expects if an axis is parallelized with DIDx then the extent of that axis == number of devices.

Reply:
It's okay to leave it as is for now.

Reply:
Could you leave a short summary of this conversation as a code comment?